File: packages/capacitor-notificare-inbox/ios/Plugin/NotificareInboxPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareInboxKit

@objc(NotificareInboxPlugin)
public class NotificareInboxPlugin: CAPPlugin {
    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.inbox().delegate = self
    }
    
    @objc func getItems(_ call: CAPPluginCall) {
        do {
            call.resolve([
                "result": try Notificare.shared.inbox().items.map { try $0.toJson() }
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func getBadge(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.inbox().badge
        ])
    }
    
    @objc func refresh(_ call: CAPPluginCall) {
        Notificare.shared.inbox().refresh()
        call.resolve()
    }
    
    @objc func open(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.inbox().open(item) { result in
            switch result {
            case let .success(notification):
                do {
                    call.resolve([
                        "result": try notification.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func markAsRead(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.inbox().markAsRead(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func markAllAsRead(_ call: CAPPluginCall) {
        Notificare.shared.inbox().markAllAsRead { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func remove(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.inbox().remove(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func clear(_ call: CAPPluginCall) {
        Notificare.shared.inbox().clear { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
}

extension NotificareInboxPlugin: NotificareInboxDelegate {
    public func notificare(_ notificareInbox: NotificareInbox, didUpdateInbox items: [NotificareInboxItem]) {
        do {
            EventBroker.instance.dispatchEvent("inbox_updated", data: [
                "items": try items.map { try $0.toJson() }
            ])
        } catch {
            NotificareLogger.error("Failed to emit the inbox_updated event.", error: error)
        }
    }
    
    public func notificare(_ notificareInbox: NotificareInbox, didUpdateBadge badge: Int) {
        EventBroker.instance.dispatchEvent("badge_updated", data: [
            "badge": badge
        ])
    }
}

File: packages/capacitor-notificare-inbox/android/src/main/java/re/notifica/inbox/capacitor/NotificareInboxPlugin.kt
package re.notifica.inbox.capacitor

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.Observer
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.inbox.ktx.inbox
import re.notifica.inbox.models.NotificareInboxItem
import re.notifica.inbox.models.fromJson
import re.notifica.inbox.models.toJson
import re.notifica.internal.NotificareLogger
import re.notifica.models.NotificareNotification
import java.util.*

@CapacitorPlugin(name = "NotificareInboxPlugin")
public class NotificareInboxPlugin : Plugin() {
    private val itemsObserver = Observer<SortedSet<NotificareInboxItem>> { items ->
        if (items == null) return@Observer

        try {
            val data = JSObject()
            data.put("items", JSArray(items.map { it.toJson() }))

            EventBroker.dispatchEvent("inbox_updated", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the inbox_updated event.", e)
        }
    }

    private val badgeObserver = Observer<Int> { badge ->
        if (badge == null) return@Observer

        EventBroker.dispatchEvent("badge_updated", JSObject().apply {
            put("badge", badge)
        })
    }

    override fun load() {
        EventBroker.setup(this::notifyListeners)

        onMainThread {
            Notificare.inbox().observableItems.removeObserver(itemsObserver)
            Notificare.inbox().observableItems.observeForever(itemsObserver)

            Notificare.inbox().observableBadge.removeObserver(badgeObserver)
            Notificare.inbox().observableBadge.observeForever(badgeObserver)
        }
    }

    // region Notificare Inbox

    @PluginMethod
    public fun getItems(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Notificare.inbox().items.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun getBadge(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.inbox().badge)
            }
        )
    }

    @PluginMethod
    public fun refresh(call: PluginCall) {
        Notificare.inbox().refresh()
        call.resolve()
    }

    @PluginMethod
    public fun open(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareInboxItem = try {
            NotificareInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.inbox().open(item, object : NotificareCallback<NotificareNotification> {
            override fun onSuccess(result: NotificareNotification) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAsRead(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareInboxItem = try {
            NotificareInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.inbox().markAsRead(item, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAllAsRead(call: PluginCall) {
        Notificare.inbox().markAllAsRead(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun remove(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareInboxItem = try {
            NotificareInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.inbox().remove(item, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clear(call: PluginCall) {
        Notificare.inbox().clear(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    public companion object {
        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}
